Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,96 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
```

Comment:
Yes, I removed it because it simplifies things and encourages the startup hook to be seen as an implicit way to inject dependencies into the process from outside at process launch, rather than a part of the app. We could add the option to runtimeconfig.json if folks need it.

I'm not sure I understand your scenario. The point of making this an environment variable was exactly so that it could be controlled by someone who controls launching of apps, but not the deployed bits. Could you clarify?

Reply:
You are correct that a 'wrapper' tool that launches the app will use the Env variable.   The issue is other scenarios.  Today apps that use AppInsights require that you modify the app to include its telemetry.   It would be a useful feature for them to be able to instrument the app for you (post build) without breaking any signing that happened (you don't want to modify their DLLs), by plopping down their DLL and modifying the app's config.    Thus there are multiple scenarios. 

I want to mull this over some more...